File: Array30/LeaderArray.js
function LeaderArray(arr) {
    const n = arr.length;
    const leaders = [];
    let maxRight = arr[n - 1]; // Initialize the maximum from the right side
  
    leaders.push(maxRight); // The last element is always a leader
  
    for (let i = n - 2; i >= 0; i--) {
      if (arr[i] > maxRight) {
        leaders.push(arr[i]);
        maxRight = arr[i]; // Update the maximum from the right side
      }
    }
  
    return leaders.reverse(); // Reverse the array to get the correct order
  }
  
  const input = [16, 17, 4, 3, 5, 2];
  const result = LeaderArray(input);
  console.log(result);
  

File: Array30/LeaderArray.py
def leader_array(arr):
    n = len(arr)
    leaders = []
    max_right = arr[n - 1]  # Initialize the maximum from the right side

    leaders.append(max_right)  # The last element is always a leader

    for i in range(n - 2, -1, -1):
        if arr[i] > max_right:
            leaders.append(arr[i])
            max_right = arr[i]  # Update the maximum from the right side

    return leaders[::-1]  # Reverse the list to get the correct order

input_list = [16, 17, 4, 3, 5, 2]
result = leader_array(input_list)
print(result)
